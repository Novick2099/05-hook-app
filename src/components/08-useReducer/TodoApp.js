import React, {useEffect, useReducer} from 'react';
import {todoReducer} from "./todoReducer";
import { Grid, Typography } from "@mui/material";
import TodoList from "./TodoList";
import TodoAdd from "./TodoAdd";

const init = ()=>{
    return JSON.parse(localStorage.getItem('todosState')) || [];
}

const TodoApp = () => {

    // la desestructuracion de arreglos es posicional
    const [todosState, dispatch] = useReducer( todoReducer, [], init );

    useEffect(() => {
        localStorage.setItem('todosState', JSON.stringify(todosState));
    }, [todosState]);

    const handleAddTodo = (newTodo) => {
        dispatch( {
            type: 'add',
            payload: newTodo,
        });
    }

    const handleDelete = ( todoId ) => {
        const action = {
            type: 'delete',
            payload: todoId,
        }
        dispatch(action);
    }

    const handleToggle = ( todoId ) => {
        dispatch({
            type: 'toggle',
            payload: todoId,
        });
    }

    return (
        <div>
            <Grid container item  spacing={5} xs={6} md={6}>
                <Grid item xs={8}>
                    <Typography variant="h2">To do List ({todosState.length})</Typography>
                    <nav aria-label="main mailGrid folders">
                        <TodoList
                          todos = {todosState}
                          handleDelete = {handleDelete}
                          handleToggle = {handleToggle}
                        />
                    </nav>
                </Grid>
                <Grid item xs={4}>
                    <TodoAdd
                      handleAddTodo = {handleAddTodo}
                    />
                </Grid>
            </Grid>
        </div>
    );
};

export default TodoApp;
