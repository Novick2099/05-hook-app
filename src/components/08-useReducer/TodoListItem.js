import React from 'react';
import IconButton from "@material-ui/core/IconButton";
import DeleteIcon from "@mui/icons-material/Delete";
import {ListItem, ListItemButton, ListItemText} from "@mui/material";

const TodoListItem = ({key, todo, index, handleDelete, handleToggle}) => {

  const onClickToggle=()=>{
    handleToggle(todo.id)
  }

  return (
    <ListItem secondaryAction={
      <IconButton id={`delete_button_${index}`} edge="end" aria-label="delete" onClick={()=>{
        handleDelete(todo.id);
      }}>
        <DeleteIcon />
      </IconButton>}
    >
      <ListItemButton
        id={`TodoToggleButton${todo.id}}`}
        key={`TodoToggleButton${todo.id}}`}
        onClick={onClickToggle}>
        <ListItemText
          key={`TodoTextItem${todo.id}}`}
          primary={`${index+1}.${todo.desc}`}
          // todo preferir usar operadores ternarios a condition && value
          sx={todo.done ? {textDecoration: 'line-through'}:{}}
        />
      </ListItemButton>
    </ListItem>
  );
};

export default TodoListItem;
