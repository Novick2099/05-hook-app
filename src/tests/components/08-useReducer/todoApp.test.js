import React from 'react';
import TodoApp from "../../../components/08-useReducer/TodoApp";
import {act} from "@testing-library/react";
import TodoListItem from "../../../components/08-useReducer/TodoListItem";
import {shallow} from "enzyme";
const {mount} = require("enzyme");

const demoTodos = [{
  id: 1,
  desc: 'Aprender React',
  done: false
},{
  id: 2,
  desc: 'Aprender React',
  done: true
}];

describe('Pruebas en TodoApp ',()=>{

  const wrapper = mount(<TodoApp />);
  const shallowTodoApp =shallow(<TodoApp />);
  Storage.prototype.setItem = jest.fn(()=>{});

  it('Should match with snapshot',()=>{
    expect(wrapper).toMatchSnapshot();
  });

  it('Sould add a todo',()=>{
    console.log(wrapper.find('TodoAdd').props());
    //todo investigar act in testing-library-react
    act(()=>{
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
    });
    act(()=>{
      wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1]);
    });
    wrapper.update();
    expect(wrapper.find(TodoListItem).length).toBe(2);
    console.log(wrapper.debug());
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);

    //Todo tip usar shallow para debuguear rapido y encontrar elementos
    // Combo
    console.log(shallowTodoApp.debug());
    expect(wrapper.text()).toContain('1');
  });
});