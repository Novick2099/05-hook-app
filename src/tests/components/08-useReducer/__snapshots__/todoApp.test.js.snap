// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoApp  Should match with snapshot 1`] = `
<div>
  Array [
    "",
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-spacing-xs-5 MuiGrid-grid-xs-6 MuiGrid-grid-md-6 css-1oxkaji-MuiGrid-root"
    >
      Array [
        "",
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-vok43g-MuiGrid-root"
        >
          Array [
            "",
            <h2
              className="MuiTypography-root MuiTypography-h2 css-1sra7t5-MuiTypography-root"
            >
              To do List (
              0
              )
            </h2>,
          ]
          <nav
            aria-label="main mailGrid folders"
          >
            <ul
              className="MuiList-root MuiList-padding"
            />
          </nav>
        </div>,
      ]
      Array [
        "",
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-gj1fbr-MuiGrid-root"
        >
          Array [
            Array [
              "",
              <h3
                className="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
              >
                Add Todo
              </h3>,
            ],
            <form
              onSubmit={[Function]}
            >
              Array [
                "",
                Array [
                  "",
                  <div
                    className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary css-a88p61-MuiInputBase-root-MuiInput-root"
                    onClick={[Function]}
                  >
                    Array [
                      "",
                      <input
                        autoComplete="off"
                        className="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                        name="description"
                        onAnimationStart={[Function]}
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        placeholder="Aprender ..."
                        type="text"
                        value=""
                      />,
                    ]
                  </div>,
                ],
              ]
              Array [
                "",
                Array [
                  "",
                  <button
                    className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
                    disabled={false}
                    onBlur={[Function]}
                    onContextMenu={[Function]}
                    onDragLeave={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    tabIndex={0}
                    type="submit"
                  >
                    Add
                    Array [
                      "",
                      <span
                        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />,
                    ]
                  </button>,
                ],
              ]
            </form>,
          ]
        </div>,
      ]
    </div>,
  ]
</div>
`;
